drivers: gpio: nrfx: extend pin retain to every pin

bjarki-andreasen opened the pull request:
The usage of nrf_gpio_port_retain_disable/_enable, in cases where the soc pins support retention, every pin must be
retained/unretained regardless of what power domain the pad is in.

This patch ensures retain is applied to all pins in all domains by the gpio_nrfx device driver, not only pins specifically in the fast_active_1 domain. Without this patch, pinctrl will correctly retain pins, while gpio_nrfx will fail to unretain them again.

bjarki-andreasen-nordic committed:
drivers: gpio: nrfx: extend pin retain to every pin

The usage of nrf_gpio_port_retain_disable/_enable, in cases where
the soc pins support retention, every pin must be
retained/unretained regardless of what power domain the pad is in.

This patch ensures retain is applied to all pins in all domains by
the gpio_nrfx device driver, not only pins specifically in the
fast_active_1 domain. Without this patch, pinctrl will correctly
retain pins, while gpio_nrfx will fail to unretain them when
again.

We no longer check the output state either, which was passed with
the flags arg of gpio_nrfx_gpd_retain_set() so this arg has been
removed.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -69,10 +69,6 @@ static void gpio_nrfx_gpd_retain_set(const struct device *port, uint32_t mask, g
 #ifdef CONFIG_SOC_NRF54H20_GPD
 	const struct gpio_nrfx_cfg *cfg = get_port_cfg(port);
 
-	if (cfg->pad_pd != NRF_GPD_FAST_ACTIVE1 || !(flags & GPIO_OUTPUT)) {
```
Since the `flags` parameter is not to be used anymore, I think the function signature should be modified accordingly.

bjarki-andreasen-nordic committed:
drivers: pinctrl: nrf: patch pin retain to follow lp mode

Pins can be configured to retain their config even when the power
domain they belong to is suspended. Update pinctrl_nrf to enable
retain only if the pin has been configured and is not in use
(pincnf is low-power/sleep), disable retain otherwise.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

karstenkoenig reviewed (approved):
Would this allow removing this workaround?
https://github.com/zephyrproject-rtos/zephyr/commit/e9eee4eb232b674aaa0f63f892cb291883abfedf

It was needed because the CAN peripheral itself sits in a fast domain, but the pins themselves don't. You are reworking this also here https://github.com/zephyrproject-rtos/zephyr/pull/90754/files by decoupling the peripheral power domain and the pad power domain but this patch here might already be enough, as retain should now be blindly disabled

bjarki-andreasen commented:
> Would this allow removing this workaround? [e9eee4e](https://github.com/zephyrproject-rtos/zephyr/commit/e9eee4eb232b674aaa0f63f892cb291883abfedf)
> 
> It was needed because the CAN peripheral itself sits in a fast domain, but the pins themselves don't. You are reworking this also here https://github.com/zephyrproject-rtos/zephyr/pull/90754/files by decoupling the peripheral power domain and the pad power domain but this patch here might already be enough, as retain should now be blindly disabled

I think it will negate the need for the can workaround yes :) but so will the larger pad rework, I'm only adding this "quick" fix to allow device_deinit to work for SPIM and TWIM and TWIS without adding workarounds to those :)
